utils/memory_bench/src/main.rs

```diff
@@ -0,0 +1,421 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/master/LICENSE ).
+use quote::quote;
+use serde_json::json;
+use std::fs::File;
+use std::io::prelude::*;
+use std::path::PathBuf;
+use std::process;
+use std::process::Command;
+use std::{env, process::Stdio};
+use std::{fs, io::BufReader};
+
+fn inject_dhat_scoped_variable(syntax: &mut syn::File) {
+    let dhat = syn::parse_str("let _dhat = Dhat::start_heap_profiling();")
+        .expect("Unable to parse the dhat string");
+
+    // Find the main function.
+    for item in syntax.items.iter_mut() {
+        if let syn::Item::Fn(ref mut fn_item) = item {
+            if fn_item.sig.ident == "main" {
+                // Create a new vector with the injected statement at the beginning.
+                let mut new_stmts = vec![dhat];
+                for stmt in fn_item.block.stmts.drain(..) {
+                    new_stmts.push(stmt);
+                }
+                fn_item.block.stmts = new_stmts;
+                return;
+            }
+        }
+    }
+
+    panic!("Unable to find the main function.");
+}
+
+fn inject_allocator_declaration(syntax: &mut syn::File) {
+    let use_code = "use dhat::{Dhat, DhatAlloc};";
+    let allocator_code = "
+        #[global_allocator]
```

Comment:
Instead of re-writing the code in this tool, could we add this boilerplate directly to the example files under a feature flag?  For example, `cargo run --features dhat --example work_log`

I think it's okay to carry boilerplate in all of the example files, but if it gets too much, we could put it in a macro that all the examples include.

Reply:
The instrumentation is not a large burden of code for each example. It's a couple lines of imports, and then a single declaration at the top of the main. I think this is a trade-off between maintaining a tool that rewrites code, and making the examples more complicated for the common case of just running an example. As a user of a project, I would appreciate the examples to be as direct as possible, as I often go there to see how code really works. It would make these examples somewhat less clear to include memory instrumentation.

I would be fine going either way depending on the feedback of the group, but my preference is to include the rewriting tool.

Reply:
I see rustdoc as the primary place to put code examples.  The example files are a secondary place to put examples, and therefore we can afford a bit of boilerplate in order to make the examples work well with tooling.

We can achieve both goals (making examples useful for testing and simple for users) by having "setup" and "teardown" macros that we include into each example.  Note also that we already have some precedent for changing example files for tooling purposes: in #365 I replaced the built-in Rust main function with `#[no_main]` in order to produce higher quality WASM files for code size testing.

Reply:
In the meeting we decided to add a macro to the examples, and remove the code rewriting.

Reply:
The new implementation relies on macros and features to enable and disable the memory instrumentation.